dbms/src/AggregateFunctions/AggregateFunctionGroupArray.h

```diff
@@ -43,51 +71,162 @@ struct GroupArrayNumericData
     Array value;
 };
 
+template <typename T>
+struct GroupArrayNumericData<T, true>
+{
+    // Switch to ordinary Allocator after 4096 bytes to avoid fragmentation and trash in Arena
+    using Allocator = MixedAlignedArenaAllocator<alignof(T), 4096>;
+    using Array = PODArray<T, 32, Allocator>;
+
+    Array value;
+    size_t total_values = 0;
+    pcg32_fast rng;
+
+    UInt64 genRandom(size_t lim)
+    {
+        /// With a large number of values, we will generate random numbers several times slower.
+        if (lim <= static_cast<UInt64>(rng.max()))
+            return static_cast<UInt32>(rng()) % static_cast<UInt32>(lim);
```

Comment:
https://github.com/lemire/fastrange to avoid division on runtime value.

Reply:
Ok, time to do some micro benchmarks :)